server/src/main/java/org/apache/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -587,6 +594,43 @@ private StringFullResponseHolder fetchLookupsForTier(String tier) throws Interru
     void handle(Map<String, LookupExtractorFactoryContainer> lookupMap) throws Exception;
   }
 
+  private static class StatusNotice implements Notice
+  {
+    private final LookupReferencesManager manager;
+    private final String lookupName;
+    private final LookupExtractorFactoryContainer lookupExtractorFactoryContainer;
+
+    public StatusNotice(
+        LookupReferencesManager manager,
+        String lookupName,
+        LookupExtractorFactoryContainer lookupExtractorFactoryContainer
+    )
+    {
+      this.manager = manager;
+      this.lookupName = lookupName;
+      this.lookupExtractorFactoryContainer = lookupExtractorFactoryContainer;
+    }
+
+    @Override
+    public void handle(Map<String, LookupExtractorFactoryContainer> lookupMap)
+    {
+      if (lookupExtractorFactoryContainer.getLookupExtractorFactory().isReady()) {
+        LookupExtractorFactoryContainer old = lookupMap.put(lookupName, lookupExtractorFactoryContainer);
+
+        LOG.info("Loaded lookup [%s] with spec [%s].", lookupName, lookupExtractorFactoryContainer);
+
+        if (old != null) {
+          if (!old.getLookupExtractorFactory().destroy()) {
+            throw new ISE("destroy method returned false for lookup [%s]:[%s]", lookupName, old);
+          }
+        }
+      } else {
+        LOG.info("Loading lookup [%s] with spec [%s].", lookupName, lookupExtractorFactoryContainer);
+        manager.addNotice(new StatusNotice(manager, lookupName, lookupExtractorFactoryContainer));
```

Comment:
For cases where a cached lookup has been "started" but the cache isn't initialized yet, what would be the behavior for a query using this lookup? Wouldn't it still fail with the CACHE_NOT_INITIALIZED error?

Reply:
If there is a new cached lookup being initialised, then this lookup will not be visible to queries. And queries will failed with `Lookup [xxx] not found` error, see https://github.com/apache/druid/blob/1b9a8c46874520d22b65573eeabfa3a62e16a5e0/processing/src/main/java/org/apache/druid/query/lookup/RegisteredLookupExtractionFn.java#L151

This PR only tries to improve the use case of updating an existing lookup with higher version. And I agree that this [proposal](https://github.com/apache/druid/issues/10294) will improve these use cases on a general level.

Reply:
Thanks.

> This PR only tries to improve the use case of updating an existing lookup with higher version.

Could you please add tests for this? It should also help with the fixing the code coverage failures in Travis.